File: mobile-app/lib/features/rule_diffs/rule_diffs_view_helpers.dart
// Pure-Dart helpers für das Mobile-rule_diffs-panel.
// Spiegelt desktop-app/rule_diffs_view_helpers.js — gleiche signatur,
// gleiche semantik. Liegt isoliert, damit ohne flutter-bindings testbar.

/// Liefert nur pending diff-Einträge aus einem project-map (wie vom server
/// kommt: dynamic JSON).
List<Map<String, dynamic>> selectPending(Map<String, dynamic>? project) {
  if (project == null) return const [];
  final raw = project['ruleDiffs'];
  if (raw is! List) return const [];
  return raw
      .whereType<Map>()
      .map((e) => e.cast<String, dynamic>())
      .where((d) => d['status'] == 'pending')
      .toList(growable: false);
}

/// Anzahl pending diffs — für badge/chip im UI.
int countPending(Map<String, dynamic>? project) => selectPending(project).length;

/// Menschen-lesbares label, z.B. „aktivieren: kein unnötiger code".
String formatDiffLabel(Map<String, dynamic>? diff) {
  if (diff == null) return '';
  final text = (diff['text'] ?? '').toString();
  switch (diff['action']) {
    case 'activate':
      return 'aktivieren: $text';
    case 'deactivate':
      return 'deaktivieren: $text';
    default:
      return 'änderung: $text';
  }
}

File: release/ProjectGamma/desktop-app/rule_diffs_view_helpers.js
// Pure helpers für die rule-diffs-UI (desktop-app).
//
// Trennt Datenformung von der jsx-Komponente, damit per node:test ohne
// browser-stack getestet werden kann. Wird sowohl in node (require) als
// auch im browser (window-global) genutzt.

(function (root, factory) {
  const mod = factory();
  if (typeof module === "object" && module.exports) {
    module.exports = mod;
  } else {
    root.RuleDiffsViewHelpers = mod;
  }
})(typeof self !== "undefined" ? self : this, function () {
  /**
   * Liefert nur pending diff-Einträge eines projects.
   * @param {{ruleDiffs?: Array}|null|undefined} project
   * @returns {Array}
   */
  function selectPending(project) {
    if (!project || !Array.isArray(project.ruleDiffs)) return [];
    return project.ruleDiffs.filter((d) => d && d.status === "pending");
  }

  /**
   * Anzahl pending diffs (für badge/chip).
   */
  function countPending(project) {
    return selectPending(project).length;
  }

  /**
   * Menschen-lesbares label für einen diff.
   */
  function formatDiffLabel(diff) {
    if (!diff) return "";
    const text = diff.text || "";
    if (diff.action === "activate") return `aktivieren: ${text}`;
    if (diff.action === "deactivate") return `deaktivieren: ${text}`;
    return `änderung: ${text}`;
  }

  return { selectPending, countPending, formatDiffLabel };
});
